packages/monaco/src/browser/monaco-outline-contribution.ts

```diff
@@ -167,43 +183,93 @@ export class MonacoOutlineContribution implements FrontendApplicationContributio
         return roots;
     }
 
-    protected parentContains(symbol: DocumentSymbol, parent: DocumentSymbol, rangeBased: boolean): boolean {
-        const symbolRange = this.getRangeFromSymbolInformation(symbol);
-        const nodeRange = this.getRangeFromSymbolInformation(parent);
-        const sameStartLine = symbolRange.start.line === nodeRange.start.line;
-        const startColGreaterOrEqual = symbolRange.start.character >= nodeRange.start.character;
-        const startLineGreater = symbolRange.start.line > nodeRange.start.line;
-        const sameEndLine = symbolRange.end.line === nodeRange.end.line;
-        const endColSmallerOrEqual = symbolRange.end.character <= nodeRange.end.character;
-        const endLineSmaller = symbolRange.end.line < nodeRange.end.line;
+    /**
+     * Sets the selection on the sub-trees based on the optional editor selection.
+     * Select the narrowest node that is strictly contains the editor selection.
+     */
+    protected applySelection(roots: MonacoOutlineSymbolInformationNode[], editorSelection?: Range): MonacoOutlineSymbolInformationNode[] {
```

Comment:
It can be optimized to avoid unnecessary iterations and respect expansion state:
```ts
    protected applySelection(roots: MonacoOutlineSymbolInformationNode[], editorSelection?: Range): boolean {
        if (editorSelection) {
            for (const root of roots) {
                if (this.parentContains(editorSelection, root.fullRange, true)) {
                    const { children } = root;
                    root.selected = !root.expanded || !this.applySelection(children, editorSelection);
                    return true;
                }
            }
        }
        return false;
    }
```